File: algorithms/quickSort.js
const swap = (nums, i, j) => ([nums[i], nums[j]] = [nums[j], nums[i]]);

const partition = (nums, loIndex, hiIndex) => {
  let pivotValue = nums[loIndex];
  let pivotIndex = loIndex;
  loIndex++;

  while (loIndex <= hiIndex) {
    if (nums[loIndex] < pivotValue) {
      loIndex++;
    } else if (nums[hiIndex] >= pivotValue) {
      hiIndex--;
    } else {
      swap(nums, loIndex, hiIndex);
    }
  }

  swap(nums, pivotIndex, hiIndex);

  return hiIndex;
};

//  O(NlogN)
const quickSort = (nums, low, high) => {
  if (low >= high) {
    return;
  }

  let pivot = partition(nums, low, high);

  quickSort(nums, low, pivot - 1);
  quickSort(nums, pivot + 1, high);
};

const input = [10, 16, 8, 12, 15, 6, 3, 9, 5];
const res = quickSort(input, 0, input.length - 1);
console.log('--', input);

File: algorithms/quickSort.py
class Solution:
    def sortArray(self, nums):
        def swap(nums, i, j):
          nums[i], nums[j] = nums[j], nums[i] 
          return nums

        def quick_select(nums, lo, hi):
          pivotIndex = lo
          pivotValue = nums[lo]
          lo += 1

          while lo <= hi:
            if nums[lo] < pivotValue:
              lo += 1
            elif nums[hi] >= pivotValue:
              hi -= 1
            else:
              swap(nums, lo, hi)

          swap(nums, pivotIndex, hi)

          return hi
        
        def quick_sort(lo, hi):
          if lo >= hi:
            return

          partition = quick_select(nums, lo, hi) # get pivot index

          quick_sort(lo, partition - 1)
          quick_sort(partition + 1, hi)

        quick_sort(0, len(nums) - 1)

        return nums

x = Solution()
x.sortArray([5,1,1,2,0,0])